### PR TITLE
cocci: Detect unlogged missed tail calls

### PR DESCRIPTION
**Pull request by pchaigno**
The first commit introduces a new Coccinelle script to detect tail calls not followed by a `DROP_MISSED_TAIL_CALL` log. The second commit adapts the code a bit in two places to allow Coccinelle to detect that all paths from tail call contain a `DROP_MISSED_TAIL_CALL` log. The last commit fixes two instances of unlogged missed tail calls (dating from Cilium 0.10.1).

@cilium/bpf Do we want to backport the two missing log for the tail calls?

Fixes: #190
Updates: #11257

### CODE REVIEW

**Comment by pchaigno** [comment]
retest-net-next

**Comment by pchaigno** [comment]
test-me-please

**Comment by pchaigno** [comment]
retest-gke

**Review by qmonnet — APPROVED** [review]
Looks great!

Looks like a candidate for at least 1.8 backport to me, in case it helps catch or trace a bug in the future? Plus it should not be difficult to backport to 1.8.

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -1303,16 +1303,17 @@ int tail_ipv6_to_ipv4(struct __ctx_buff *ctx)
...
 	cilium_dbg_capture(ctx, DBG_CAPTURE_AFTER_V64, ctx->ingress_ifindex);
 
 	ctx_store_meta(ctx, CB_NAT46_STATE, NAT64);
 
 	invoke_tailcall_if(__and(is_defined(ENABLE_IPV4), is_defined(ENABLE_IPV6)),
 			   CILIUM_CALL_IPV4_FROM_LXC, tail_handle_ipv4);
-	return ret;
+drop_err:
+	return send_drop_notify(ctx, SECLABEL, 0, 0, ret, CTX_ACT_DROP,
+				METRIC_EGRESS);
```
Comment (borkmann) [inline bpf/bpf_lxc.c:1316]: I don't think this is correct.

Comment (borkmann) [reply bpf/bpf_lxc.c:1316]: `invoke_tailcall_if ()` results into a normal/inlined call if the v4/v6 condition is not satisfied, so this means we always send a drop notification for these cases, no?

Comment (pchaigno) [reply bpf/bpf_lxc.c:1316]: @borkmann If the conditions are not met, it returns as far as I can see, so we won't hit the `send_drop_notify()`:
https://github.com/cilium/cilium/blob/1dc3e396c1e048c94ad65d6f6bdb0463bf3281fb/bpf/lib/tailcall.h#L90-L91

Comment (borkmann) [reply bpf/bpf_lxc.c:1316]: Good point, you're right! Merging then. 👍

**Review by borkmann — APPROVED** [review]

**Pull request merged**